rep-0151.rst

```diff
@@ -0,0 +1,151 @@
+REP: 151
+Title: Python 3 Support
+Author: Dirk Thomas
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 23-Dec-2017
+
+Outline
+=======
+
+#. Abstract_
+#. Motivation_
+#. Rationale_
+#. Proposal_
+#. Requirements_
+#. References_
+#. Copyright_
+
+Abstract
+========
+
+This REP describes the path for switching from Python 2 to Python 3 in ROS 1.
+
+Motivation
+==========
+
+As described in the Python release schedule [1]_ Python 2.7 will only receive
+bugfix releases until 2020.
+As a consequence Ubuntu is trying to demote Python 2.7 into the `universe`
+repository for 18.04 since they don't want to guarantee support for 5 years.
+While it is unlikely that Python 2.7 will be removed from the archives entirely
+it should not be relied on past its EOL data.
+Therefore ROS 1 needs to begin moving to Python 3 in order to make the
+transition as smooth as possible.
+
+Rationale
+=========
+
+The goal for a future ROS 1 distribution to support Python 3 should be achieved
+no later than for ROS O-turtle (Mai 2020).
+Since O-turtle is a LTS release it is strongly desired to already support
+Python 3 in the ROS N-turtle release to allow enough testing time before the
+mandatory switch.
+
+In general a single Python version must be used since different versions can't
+be mixed within a single ROS distribution.
+E.g. a package A using Python 3 can't use the Python modules from package B if
+that was built with/for Python 2.
+
+In order to support Python 3 it is mandatory that the testing infrastructure
+(namely `devel` and `PR` jobs) can run the build and tests with Python 3.
+Since each of these jobs require Debian packages of the dependencies to be
+present being able to test for a specific Python version requires Debian
+packages for this Python version to be present.
+The alternative of building all dependencies from-source in these jobs doesn't
+scale and is therefore not feasible since the time necessary for a single build
+gets increasingly high the further up in the stack the package is.
+
+Updating all ROS packages which contain Python code to support Python 2
+as well as 3 is a significant effort.
+Therefore it is expected that this will take a while and will require the ROS
+community to share the necessary workload.
+
+Proposal
+========
+
+For a smooth transition the following timeline is being proposed:
+
+ * ROS O-turtle (May 2020, LTS):
+
+   * The only Debian packages provided are using Python 3.
+   * From-source builds are using Python 3 by default.
+   * All ROS packages should support Python 3.
```

Comment:
I suggest: "All actively maintained ROS packages"

Clearly, there will be may packages that never get upgraded.

Reply:
I completely agree with you that likely "some" packages will never be upgraded. But I don't see "All actively maintained ROS packages" as a good phrase either.  I expect "some" unmaintained packages to be upgraded by random people in a one-time effort and still be unmaintained afterwards.

Reply:
Agreed. Perhaps "still useful" might be a better qualifier?

Reply:
"Still useful" seems to be too fuzzy in my opinion. I don't think this needs any adjective to identify a subset of packages. All packages which want to be compatible with this ROS distribution will need to comply with it.

Reply:
Not too fuzzy: it's the fact that someone still wants to use a package that triggers the effort involved in porting and testing. Otherwise, who cares?

Reply:
I think since we're in the scope of O-turtle, we don't need to qualify the statement "All ROS packages should support Python 3." If you're unmaintained/unreleased you're not in O-turtle, if you are released/maintained you must support python 3.